### DISCUSSION TITLE
continuous time-dynamic network embeddings - a question about random walks and large communities

### DISCUSSION THREAD

**Discussion by amjass12** [stellargraph/stellargraph#1815] [General]
Hello!

I have a question regarding some rules re- random walks. The word2vec method of temporal embeddings is extremely useful (i previously posted about temporal graph_net models) but feel this method best suits my application as it preserves node and community structure in an elegant manner (both temporally but also in individual graphs when running static walks and carrying out embeddings in word2vec). The communities in each temporal graph (or single graph) are very large.. there are some communities that consist of >300 nodes. I therefore wonder if my logic in increasing walk length is correct. right now i am using default tutorial parameters:

`walk_length = 80`
Question: should this walk length be increased in graphs with much larger communities to create a richer representation of the node vectors that will be fed in to the word2vec model? if I do this, should num_walks_per_node also increase? (presumably these parameters are also a 'see which works best approach')?

**weighted graphs**:  all of my graphs are weighted and the edge weights between nodes matter.  the graph is constructed as follows from networkx:

`graph = StellarGraph.from_networkx(networkGraph12, edge_weight_attr = 'value')` (presumably `node_features` are not critical here)

question:for single static graphs (for downstream applications and time dynamic word embeddings), would a weighted random walk improve the word embedding in word2vec if weighted random walks are carried out as opposed to a `BiasedRandomWalk`

many thanks for your time!

**Comment by adocherty** [comment]
Hi, thanks for the questions!  These questions are difficult to answer, typically as you correctly pointed out the "see which works best" approach is the most straightforward and any generalizations come after seeing what approach has worked best in different graphs.

The length of the walk will not affect either model much as the walks only depend on the previous node - the walk length here is just for computational efficiency (see paper [1] for details).  However, the `context_window_size` parameter could be useful: making it larger will make the model consider pairs of nodes further apart as being similar.

For static graphs the `BiasedRandomWalk` is the implementation of Nove2Vec – the 2nd order walk described in the paper [1] – which introduces new parameters `p` and `q` where the return parameter, `p`, controls the likelihood of immediately revisiting a node in the walk and the in-out parameter `q` allows the search to differentiate between “inward” and “outward” nodes.  The paper proposed that these roughly map to emphasizing nodes that are organized based on communities they belong to (i.e.,homophily) versus nodes that are organized based on the structural roles of nodes in the  network  (i.e.,structural equivalence).

These parameters would be useful for you to analyze larger communities by setting the parameters to favor the walks exploring further into the graph. Note that for temporal walks these parameters are not available as the walk is always one way - each step increasing in time.

Whether this is better for you than a weighted random walk is very much dependent upon your problem and graph – and will have to be assessed by a "see which works best" approach.

[1] Node2Vec: Scalable Feature Learning for Networks. A. Grover, J. Leskovec. ACM SIGKDD International Conference on Knowledge Discovery and Data Mining (KDD), 2016. (link)
